lib/sync/replication-state.js

```diff
@@ -0,0 +1,279 @@
+import { truncate } from 'fs'
+import { TypedEmitter } from 'tiny-typed-emitter'
+import { idToKey, keyToId, truncateId } from '../utils.js'
+
+export class ReplicationState extends TypedEmitter {
+  #coreStates = new Map()
+
+  /**
+   * @param {Object} options
+   * @param {import('../core-manager/index.js').Namespace} options.namespace
+   * @param {import('../core-manager/index.js').CoreManager} options.coreManager
+   */
+  constructor(options) {
+    super()
+    this.coreManager = options.coreManager
+    this.namespace = options.namespace
+
+    const coreRecords = this.coreManager.getCores(this.namespace)
+    for (const coreRecord of coreRecords) {
+      this.#core(coreRecord)
+    }
+
+    this.coreManager.on('add-core', (coreRecord) => {
+      this.#core(coreRecord)
+    })
+  }
+
+  get state () {
+    const cores = [...this.#coreStates.entries()]
+      .reduce((obj, [ corePublicId, state ]) => {
+        obj[corePublicId] = state.state
+        return obj
+      }, {})
+
+    const synced = this.isSynced()
+
+    return { cores, synced }
+  }
+
+  get peers () {
+    const peers = new Set()
+
+    for (const state of this.#coreStates.values()) {
+        for (const peer of state.peers) {
+            peers.add(peer)
+        }
+    }
+
+    return [...peers]
+  }
+
+  isSynced () {
+    for (const state of this.#coreStates.values()) {
+        if (!state.isSynced()) {
+            return false
+        }
+    }
+
+    return true
+  }
+
+  close() {
+    this.removeAllListeners()
+    for (const state of this.#coreStates.values()) {
+      state.close()
+    }
+  }
+
+  /**
+   *
+   * @param {import('../core-manager/core-index.js').CoreRecord} coreRecord
+   */
+  async #core(coreRecord) {
+    const core = coreRecord.core
+    await core.ready()
+    const corePublicId = keyToId(coreRecord.key)
+
+    if (this.#coreStates.has(corePublicId)) {
+      return this.#coreStates.get(corePublicId)
+    }
+
+    const state = new CoreReplicationState({
+      core,
+    })
+
+    state.on('state', () => {
+      this.emit('state', this.state)
+    })
+
+    state.on('synced', () => {
+      if (this.isSynced()) {
+        this.emit('synced')
+      }
+    })
+
+    this.#coreStates.set(corePublicId, state)
+  }
+}
+
+/**
+ * @typedef {Object} PeerState
+ * @property {Number} want
+ * @private
+ */
+
+
+export class CoreReplicationState extends TypedEmitter {
+  /** @type {import('hypercore')} */
+  #core
+
+  /** @type {Map<PeerPublicId, PeerState>} */
+  #state = new Map()
+
+  /**
+   * @param {Object} options
+   * @param {import('hypercore')} options.core
+   */
+  constructor(options) {
+    super()
+    this.#core = options.core
+    this.#setInitialCoreState(this.#core)
+
+    for (const peer of this.#core.peers) {
+      this.#onPeer(peer)
+    }
+
+    this.#core.on('peer-add', (peer) => {
```

Comment:
I _think_ it'll have the same lifecycle as the app.

Oh, or in multi-projects land it might be equivalent to the lifecycle of the active project.

Unless we have a sync party mode that syncs multiple projects at the same time.

I thought I had an answer to this and then quickly realized I didn't, haha.